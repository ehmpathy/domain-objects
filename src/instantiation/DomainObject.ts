import { withImmute, WithImmute } from '../manipulation/immute/withImmute';
import { DomainObjectShape } from '../reference/DomainReferenceable';
import { hydrateNestedDomainObjects } from './hydrate/hydrateNestedDomainObjects';
import { SchemaOptions, validate } from './validate/validate';

export interface DomainObjectInstantiationOptions {
  /**
   * allow callers to skip certain aspects of instantiation
   *
   * e.g., for performance optimizations in specific usecases
   */
  skip?: {
    /**
     * allow callers to skip schema validation
     *
     * usecase examples
     * - deserialize: schema validation increases deserialization time dramatically (10-100x) (especially if using Joi, that thing is slow!)
     */
    schema?: boolean;
  };
}

/**
 * Domain Object
 *
 * Responsibilities:
 * - optionally validate the properties that are passed into the constructor against the schema at runtime, if schema is supplied
 * - assign all properties that are passed into constructor to self, after optional runtime validation
 */
export class DomainObject<T extends DomainObjectShape> {
  constructor(props: T, options?: DomainObjectInstantiationOptions) {
    // 1. validate with the schema if provided
    const { schema } = this.constructor as typeof DomainObject; // `this.constructor` does not get typed to DomainObject automatically by ts; https://stackoverflow.com/questions/33387318/access-to-static-properties-via-this-constructor-in-typescript
    if (schema && !(options?.skip?.schema === true))
      validate({ props, schema, domainObjectName: this.constructor.name });

    // 2. hydrate any nested props present; just overwrite the orig props for each "nested" key
    const nested = ((this.constructor as typeof DomainObject).nested ??
      {}) as Record<string, typeof DomainObject>;
    const hydratedProps = hydrateNestedDomainObjects({
      domainObjectName: this.constructor.name,
      props,
      nested,
    });

    // 3. assign all properties to self if passed validation
    Object.assign(this, hydratedProps);
  }

  /**
   * DomainObject.alias
   *
   * When set, declares the alias via which domain objects of this class are known to be referred to.
   *
   * For example
   * - PlantPot.alias = 'pot'
   * - Bird.alias = { singular: 'bird', plural: 'flock' }
   *
   * Relevance
   * - formalizes that we can expect to see the dobj referred to by this alias
   * - enables code generators to use this colloquial alias in contracts for better devex
   */
  public static alias?: string | { plural?: string; singular?: string };

  /**
   * DomainObject.schema
   *
   * When set, will be used to validate the properties passed into the constructor at runtime (i.e., during instantiation)
   *
   * Supports [`Zod`](https://github.com/colinhacks/zod), [`Joi`](https://github.com/sideway/joi), and [`Yup`](https://github.com/jquense/yup)
   */
  public static schema?: SchemaOptions<any>; // todo: work around typescript's "static members cant reference class type parameters"

  /**
   * DomainObject.nested
   *
   * When set, will be used to hydrate nested DomainObjects. This is especially useful when instantiating nested domain objects from data stores (e.g., apis or databases). This is _required_ in order to safely use the `serialize` and `getUniqueIdentifier` functions - as those functions check static properties of domain objects to function correctly.
   *
   * For example:
   * ```ts
   * // define the plant pot
   * interface PlantPot {
   *   diameterInInches: number;
   * }
   * class PlantPot extends DomainObject<PlantPot> implements PlantPot {}
   *
   * // define the plant owner
   * interface PlantOwner {
   *   diameterInInches: number;
   * }
   * class PlantOwner extends DomainObject<PlantOwner> implements PlantOwner {}
   *
   * // define the plant
   * interface Plant {
   *   pot: PlantPot;
   *   owners: PlantOwner[];
   *   lastWatered: string;
   * }
   * class Plant extends DomainObject<Plant> implements Plant {
   *   public static nested = { pot: PlantPot, owners: PlantOwner }
   * }
   *
   * // and show that it hydrates the plant pot
   * const plant = new Plant({ pot: { diameterInInches: 7 }, owners: [{ name: 'bob' }], lastWatered: 'monday' }); // notice how the param of `pot` we're inputting is _not_ an instance of `PlantPot`, and `owners` are not instances of `PlantOwner`
   * expect(plant.pot).toBeInstanceOf(PlantPot); // now succeeds, as during instantiation we hydrated `pot` into an instance of `PlantPot`, due to `Plant.nested.pot` being set
   * plant.owners.forEach((owner) => expect(owner).toBeInstance(PlantOwner)); // also succeeds, since during instantiation we hydrated each `owner` into an instance of `PlantOwner`, due to `Plant.nested.owners` being set
   * ```
   */
  public static nested?: Record<string, DomainObject<any>>; // TODO: find a way to make this Record<keyof string>

  /**
   * DomainObject.metadata
   *
   * When set, customizes the keys that are considered as metadata of this domain object.
   *
   * Context,
   * - domain objects are often persisted inside of storage mechanisms that assign metadata to them, such as ids or timestamps
   * - metadata simply adds information _about_ the object, without contributing to _defining_ the object
   *
   * Relevance,
   * - metadata properties do not contribute to the unique key of a DomainLiteral
   * - metadata properties can be easily stripped from an object by using the `omitMetadataValues` method
   *
   * By default,
   * - `id`, `createdAt`, `updatedAt`, and `effectiveAt` are considered metadata keys
   * - `uuid` is also considered a metadata key, if it is not included in the unique key of the DomainEntity or DomainEvent
   */
  public static metadata: readonly string[];

  /**
   * .what = an interface via which to construct instances w/ immute operations
   * .why =
   *   - immute operations such as .clone produce more maintainable code by preventing unexpected mutations
   *   - these immute operations provide a safe pit of success for common operations
   * .note =
   *   - you can add withImmute to any dobj yourself, even if it wasn't built via this .build procedure
   *   - you can override the .build to add your own domain's getters, too
   */
  static build<TInstance extends DomainObjectShape>(
    this: new (props: TInstance) => TInstance,
    props: ConstructorParameters<typeof this>[0],
  ): WithImmute<TInstance> {
    const instance = new this(props);
    return withImmute(instance);
  }
}
