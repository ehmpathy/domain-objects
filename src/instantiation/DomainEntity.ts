import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, an Entity is a type of Domain Object for which:
 * - properties change over time (e.g., it has a life cycle)
 * - identity matters (e.g., two entities could have the same properties, differing only by id, and still be valid)
 *
 * For example,
 * - A `User { uuid, name, address, profilePicture }` is an entity. Properties like `name`, `address`, and `profilePicture` can change over time.
 * - A `Job { uuid, requestId, status }` is an entity. We expect `status` to change over time as the job goes from, for example, `QUEUED` to `FULFILLED`.
 */
export class DomainEntity<T> extends DomainObject<T> {
  /**
   * `DomainEntity.unique` defines all of the properties of the entity that the entity is naturally unique on
   *
   * For example,
   * - a `FileProcessingJob { uuid, filePath, status }` is likely going to be unique on the `filePath` if we only ever need to process the same file once.
   * - a `SupportTicket { uuid, userId, type, status }`, on the other hand, is likely only going to be unique on the `uuid`, because the same user can create the same support ticket over and over again
   */
  public static unique: readonly string[];

  /**
   * `DomainEntity.updatable` defines all of the properties of the entity that are updatable
   *
   * For example,
   * - a `User { uuid, name, address, picture }` would likely have `unique = ['name', 'address', 'picture']`, to enable the user to update their information as it changes over time
   * - an `Email { uuid, externalId, toAddress, fromAddress, message, status }` is likely only going to have `status` be updatable, as by changing the `toAddress` would, in this case, make us consider it to be a different email!
   */
  public static updatable: readonly string[];
}
