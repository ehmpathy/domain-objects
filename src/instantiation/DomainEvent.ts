import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, an Event is a type of Domain Object for which:
 * - properties are immutable (e.g., something happened)
 * - identity does matter
 *   - i.e., it is uniquely identified by what happened and to what
 *   - i.e., if you change the uniquely identifying properties (when or to what) of an event, it will be considered a new event
 *   - i.e., not all properties on the event uniquely identify the event
 *
 * The purpose of a Domain Event is to represent interesting things that happen in the domain.
 *
 * For example,
 * - An `ItemViewedEvent { itemUuid, customerUuid, occurredAt }` is an event. Changing when it `occurredAt` which `customerUuid` did it or which `itemUuid` it happened to would result in a new event.
 * - A `PowerSupplyMeasuredEvent { batteryUuid, current, voltage, temperature, occurredAt }` is an event. Changing when it `occurredAt` or which `batteryUuid` it occurred to would result in a new event. Note: changing the `current`, `voltage`, or `temperature` would _not_ result in a new event though, since only one measurement can be made at a time for a `batteryUuid`
 */
export abstract class DomainEvent<
  T extends Record<string, any>,
> extends DomainObject<T> {
  /**
   * `DomainEvent.unique` defines all of the properties of the event that the event is naturally unique on.
   *
   * It should always include a timestamp.
   *
   * For example,
   * - a `FileOpenedEvent { fileUuid, userUuid, occurredAt }` is likely going to be unique on the all of the keys: the combination of `['fileUuid', 'userUuid', 'occurredAt']` uniquely identifies the event.
   * - an `AirComfortMeasuredEvent { locationUuid, sensorUuid, temperature, humidity, occurredAt }`, on the other hand, is likely only going to be unique on `['locationUuid', 'sensorUuid', 'occurredAt']`, because logically we cant have a measurement from the same `sensor` + `location` at the same `time`
   */
  public static unique: readonly string[];
}
