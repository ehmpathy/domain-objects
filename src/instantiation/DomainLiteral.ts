import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, a Literal (a.k.a. Value Object), is a type of Domain Object for which:
 * - properties are immutable
 *   - i.e., it represents some literal value which happens to have a structured object shape
 *   - i.e., if you change the value of any of its properties, it is a different literal
 * - identity does not matter
 *   - i.e., it is uniquely identifiable by its non-metadata properties
 *
 * The purpose of a Domain Literal is to represent distinct, literal values which happen to have a structured object shape
 *
 * For example,
 * - A `Address { street, city, state, country }` is a literal. Changing any of the properties, like `street`, produces a completely new address
 * - A `Geocode { latitude, longitude }` is a literal. Changing either property means you are dealing with a new geocode
 */
export abstract class DomainLiteral<
  T extends Record<string, any>,
> extends DomainObject<T> {}
