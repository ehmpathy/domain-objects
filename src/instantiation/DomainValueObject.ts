import { DomainObject } from './DomainObject';

/**
 * In Domain Driven Design, a Value Object is a type of Domain Object for which:
 * - properties are immutable (e.g., its a fact)
 * - identity does not matter
 *   - i.e., it is uniquely identifiable by its properties
 *   - i.e., if you change the value of any of its properties, it is now considered a new value-object
 *
 * For example,
 * - A `Address { street, city, state, country }` is a value object. Changing any of the properties, like `street`, produces a completely new address
 * - A `Name { first, last }` is a value object. Changing either property means you are dealing with a new name.
 */
export abstract class DomainValueObject<T> extends DomainObject<T> {}
