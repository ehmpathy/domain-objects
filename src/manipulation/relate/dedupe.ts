import { DomainEntity } from '../../instantiation/DomainEntity';
import { DomainObject } from '../../instantiation/DomainObject';
import { UnexpectedCodePathError } from '../../utils/errors/UnexpectedCodePathError';
import { getUniqueIdentifier } from '../getUniqueIdentifier';
import { omitMetadataValues } from '../omitMetadataValues';
import { serialize } from '../serde/serialize';

export const dedupe = <T>(
  objs: T[],
  options?: {
    /**
     * specifies what to do if more than one version of the same entity is found
     *
     * options
     * - FAIL_FAST = throw an UnexpectedCodePathError to fail fast and prevent accidental data loss
     * - CHOOSE_FIRST_OCCURRENCE = choose the first version of this entity found in the array
     *
     * note
     * - default = FAIL_FAST, to create a pit-of-success
     */
    onMultipleEntityVersions: 'FAIL_FAST' | 'CHOOSE_FIRST_OCCURRENCE';
  },
): T[] =>
  objs.filter((thisObj, thisIndex) => {
    // determine whether this is the first occurrence of this dobj in the array
    const indexOfFirstOccurrence = objs.findIndex(
      (otherObj) =>
        serialize(
          thisObj instanceof DomainObject
            ? getUniqueIdentifier(thisObj)
            : thisObj,
        ) ===
        serialize(
          otherObj instanceof DomainObject
            ? getUniqueIdentifier(otherObj)
            : otherObj,
        ),
    );
    const isFirstOccurrence = indexOfFirstOccurrence === thisIndex;

    // if this dobj is the first occurrence, then defo not a dupe
    if (isFirstOccurrence) return true;

    // if this dobj is not the first occurrence and it is an entity, then sanity check that there are no changes between the updatable attributes
    if (
      thisObj instanceof DomainEntity &&
      options?.onMultipleEntityVersions !== 'CHOOSE_FIRST_OCCURRENCE' // if they didn't explicitly ask to choose first occurrence, then check for versions
    ) {
      const firstOccurrence = objs[indexOfFirstOccurrence];
      const foundDifferentAttributes =
        serialize(
          firstOccurrence instanceof DomainObject
            ? omitMetadataValues(firstOccurrence)
            : firstOccurrence,
        ) !==
        serialize(
          thisObj instanceof DomainObject
            ? omitMetadataValues(thisObj)
            : thisObj,
        );
      if (foundDifferentAttributes)
        throw new UnexpectedCodePathError(
          `More than one version of the same entity found in the array. Can not safely dedupe, since we don't know which version should be kept.`,
          {
            firstOccurrence,
            nextOccurrence: thisObj,
          },
        );
    }

    // otherwise, this is a dupe, and should be removed
    return false;
  });
