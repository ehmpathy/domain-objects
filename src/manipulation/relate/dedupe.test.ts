import { getError } from 'test-fns';

import { DomainEntity } from '../../instantiation/DomainEntity';
import { DomainLiteral } from '../../instantiation/DomainLiteral';
import { dedupe } from './dedupe';

describe('dedupe', () => {
  interface SeaTurtle {
    id?: number;
    saltwaterSecurityNumber: string;
    name: string;
  }
  class SeaTurtle extends DomainEntity<SeaTurtle> implements SeaTurtle {
    public static unique = ['saltwaterSecurityNumber'];
  }

  interface Region {
    name: string;
  }
  class Region extends DomainLiteral<Region> implements Region {}

  // some entities with a dupe
  const turtleOne = new SeaTurtle({
    saltwaterSecurityNumber: '821',
    name: 'Shelly C.',
  });
  const turtleTwo = new SeaTurtle({
    saltwaterSecurityNumber: '921',
    name: 'Shellbert',
  });
  const turtleOneWithId = new SeaTurtle({
    ...turtleOne,
    id: 821,
  });

  // some literals with a dupe
  const regionOne = new Region({
    name: 'Great Barrier Reef',
  });
  const regionTwo = new Region({
    name: 'Puerto Rico Trench',
  });
  const regionThree = new Region({
    name: 'East Pacific',
  });

  it('should keep only the one instance per distinct identity', () => {
    const deduped = dedupe([
      // entities
      turtleOne,
      turtleTwo,
      turtleTwo, // exact dupe
      turtleOneWithId, // identity dupe
      turtleTwo, // exact dupe
      regionOne,
      regionThree,
      regionTwo,
      regionTwo,
      regionOne,
    ]);
    expect(deduped).toHaveLength(5);
  });
  it('should keep only the first instance per distinct identity', () => {
    const deduped = dedupe([
      turtleOne,
      turtleOneWithId, // identity dupe
      turtleOne, // exact dupe
      turtleOneWithId, // identity dupe
    ]);
    expect(deduped).toHaveLength(1);
    expect(deduped[0].id).toEqual(undefined);
  });
  it('should fail fast with a helpful error if there are multiple versions of the same entity, by default', () => {
    const turtleOneWithDiffName = new SeaTurtle({
      ...turtleOne,
      name: 'Shellina C.',
    });
    const error = getError(() => dedupe([turtleOne, turtleOneWithDiffName]));
    expect(error.message).toContain(
      'More than one version of the same entity found in the array',
    );
    expect(error.message).toMatchSnapshot();
  });
  it('should choose the first version if are multiple versions of the same entity and the caller specified this mode', () => {
    const turtleOneWithDiffName = new SeaTurtle({
      ...turtleOne,
      name: 'Shellina C.',
    });

    const deduped = dedupe(
      [
        turtleOne,
        turtleOneWithDiffName, // identity dupe
        turtleOne, // exact dupe
        turtleOneWithDiffName, // identity dupe
      ],
      { onMultipleEntityVersions: 'CHOOSE_FIRST_OCCURRENCE' },
    );
    expect(deduped).toHaveLength(1);
    expect(deduped[0].name).toEqual('Shelly C.');
  });
});
