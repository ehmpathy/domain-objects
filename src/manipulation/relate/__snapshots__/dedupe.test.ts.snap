// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dedupe should fail fast with a helpful error if there are multiple versions of the same entity, by default 1`] = `
"BadRequestError: Two different versions of the same entity were asked to be deduped. Options.onMultipleEntityVersions !== 'CHOOSE_FIRST_OCCURRENCE', so we're failing fast here, since we don't know which version should be kept.

{
  "thisObj": {
    "saltwaterSecurityNumber": "821",
    "name": "Shellina C."
  },
  "versionCurrSeen": "{\\"_dobj\\":\\"SeaTurtle\\",\\"name\\":\\"Shellina C.\\",\\"saltwaterSecurityNumber\\":\\"821\\"}",
  "versionPrevSeen": "{\\"_dobj\\":\\"SeaTurtle\\",\\"name\\":\\"Shelly C.\\",\\"saltwaterSecurityNumber\\":\\"821\\"}"
}"
`;
